File: group-anagrams/obzva.py
'''
풀이
- 두 단어가 anagram 관계라면 각 단어의 알파벳 수를 계산한 결과가 같을 것입니다
- get_key 함수를 이용해서 단어를 이루는 알파벳 수에 따라 고유한 key를 생성합니다
- key를 관리하는 해시맵인 key_map을 이용하여 현재 바라보고 있는 단어와 anagram인 단어가 있는지 확인합니다

Big O
- N: 배열 strs의 크기
- K: 배열 strs의 원소 중 가장 길이가 긴 문자열의 길이

- Time complexity: O(N * K)
  - 배열 strs를 순회합니다 -> N
  - 각 문자열마다 알파벳의 수를 세기 위해 한 번 순회합니다 -> K

- Space complexity: O(N)
  - 배열 strs의 원소 모두 고유한 key를 지니고 있을 수 있습니다
    이 경우 key_map의 크기는 N에 비례하여 선형적으로 증가할 수 있습니다 -> N
'''

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        def get_key(s: str) -> str:
            count = [0] * 26
            for char in s:
                count[ord(char) - ord('a')] += 1
            res = ""
            for c in count:
                res += str(c) + ","
            return res

        idx = 0
        key_map = {}
        res = []
        for s in strs:
            key = get_key(s)
            if key not in key_map:
                key_map[key] = idx
                idx += 1
                res.append([])
            res[key_map[key]].append(s)

        return res

File: word-search/obzva.cpp
/**
 * 풀이
 * - dfs와 backtracking을 이용하여 풀었습니다
 * 
 * Big-O
 * - M: 주어진 grid `board`의 행 수
 * - N: 주어진 grid `board`의 열 수
 * - W: 주어진 string `word`의 size
 * 
 * - Time complexity: O(M * N * 3 ^ W)
 *   - `exist`함수가 grid 원소 모두를 조회합니다 -> O(M * N)
 *   - 만약 `dfs`함수가 실행될 경우, 해당 함수는 최대 3방향에 대해 재귀호출을 실행합니다 (이전 좌표로는 `dfs`를 호출하지 않기 때문)
 *   - 재귀 호출 스택의 크기는 주어진 string `word`의 길이에 비례합니다 -> O(3^W)
 *   
 * - Space complexity: O(M * N + W)
 *   - 재귀 호출 스택의 크기는 주어진 string `word`의 길이에 비례합니다 -> O(W)
 *   - 탐색 여부를 기록하는 `visit` 배열의 크기는 `board`와 같습니다 -> O(M * N)
 */

class Solution {
public:
    bool dfs(vector<vector<char>>& board, string word, vector<vector<bool>>& visit, int idx, int r, int c) {
        if (word.size() - 1 == idx) return true;

        pair<int, int> dirs[4] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int R = board.size();
        int C = board[0].size();

        visit[r][c] = true;

        int next_idx = idx + 1;
        
        bool res = false;

        for (auto dir : dirs) {
            int next_r = r + dir.first;
            int next_c = c + dir.second;

            if (0 <= next_r && next_r < R && 0 <= next_c && next_c < C && !visit[next_r][next_c]) {
                if (board[next_r][next_c] == word[next_idx] && dfs(board, word, visit, next_idx, next_r, next_c)) {
                    res = true;
                    break;
                }
            }
        }

        visit[r][c] = false;

        return res;
    }

    bool exist(vector<vector<char>>& board, string word) {
        int R = board.size();
        int C = board[0].size();
        vector<vector<bool>> visit;
        for (int i = 0; i < R; i++) {
            vector<bool> tmp;
            for (int j = 0; j < C; j++) {
                tmp.push_back(false);
            }
            visit.push_back(tmp);
        }

        for (int i = 0; i < R; i++) {
            for (int j = 0; j < C; j++) {
                if (board[i][j] == word[0] && dfs(board, word, visit, 0, i, j)) return true;
            }
        }

        return false;
    }
};
